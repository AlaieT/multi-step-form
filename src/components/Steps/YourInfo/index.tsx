import React from "react";

import Input from "../../Input";
import Button from "../../Button";
import Wrap from "../Wrap";

import type { YourInfoProps } from "../../../types";

import styles from "../../../styles/components/steps/yourInfo.module.scss";

const YourInfo = ({
  trigger,
  register,
  getValues,
  errors,
  changeStep
}: YourInfoProps) => {
  const handleNewStep = async () => {
    if (await trigger(["name", "email", "phone"])) changeStep(1);
  };

  return (
    <Wrap>
      <div id={styles.content}>
        <div id={styles.header}>
          <h1>Personal info</h1>
          <h2>Please provide your name, email address, and phone number.</h2>
        </div>
        <div id={styles.fields}>
          <Input
            {...register("name")}
            className={styles.field}
            id="name"
            label="Name"
            type="text"
            error={errors.name?.message}
            placeholder="e.g. Stephen King"
            defaultValue={getValues("name")}
          />
          <Input
            {...register("email")}
            className={styles.field}
            id="email"
            label="Email Address"
            type="email"
            error={errors.email?.message}
            placeholder="e.g. stephenking@lorem.com"
            defaultValue={getValues("email")}
          />
          <Input
            {...register("phone")}
            className={styles.field}
            id="phone-number"
            type="tel"
            label="Phone Number"
            error={errors.phone?.message}
            placeholder="e.q. +1 234 567 890"
            defaultValue={getValues("phone")}
          />
        </div>
      </div>
      <div id={styles.controls}>
        <Button
          id={styles.next_step}
          type="button"
          onClick={() => handleNewStep()}
        >
          Next Step
        </Button>
      </div>
    </Wrap>
  );
};

export default YourInfo;
