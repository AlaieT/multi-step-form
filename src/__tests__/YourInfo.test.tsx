import React from "react";
import { useForm } from "react-hook-form";
import {
  render,
  cleanup,
  screen,
  fireEvent,
  renderHook,
  act
} from "@testing-library/react";

import YourInfo from "../components/Steps/YourInfo";

afterEach(cleanup);

describe("YourInfo", () => {
  const mockChangeStep = jest.fn();

  describe("correct render", () => {
    it("should match snapshot", () => {
      const { result } = renderHook(() => useForm(global.getUseFormProps()));

      expect(
        render(
          <YourInfo
            trigger={result.current.trigger}
            register={result.current.register}
            getValues={result.current.getValues}
            errors={result.current.formState.errors}
            changeStep={mockChangeStep}
          />
        ).asFragment()
      ).toMatchSnapshot();
    });
  });

  describe("correct functionality", () => {
    it("should does correct input with no errors", async () => {
      const { result } = renderHook(() => useForm(global.getUseFormProps()));

      render(
        <YourInfo
          trigger={result.current.trigger}
          register={result.current.register}
          getValues={result.current.getValues}
          errors={result.current.formState.errors}
          changeStep={mockChangeStep}
        />
      );

      fireEvent.input(screen.getByRole("textbox", { name: "Name" }), {
        target: { value: "Some Fancy-name" }
      });
      fireEvent.input(screen.getByRole("textbox", { name: "Email Address" }), {
        target: { value: "some@mail.dom" }
      });
      fireEvent.input(screen.getByRole("textbox", { name: "Phone Number" }), {
        target: { value: "+1 555 555 5555" }
      });

      expect(screen.getByRole("textbox", { name: "Name" })).toHaveValue(
        "Some Fancy-name"
      );
      expect(
        screen.getByRole("textbox", { name: "Email Address" })
      ).toHaveValue("some@mail.dom");
      expect(screen.getByRole("textbox", { name: "Phone Number" })).toHaveValue(
        "+1 555 555 5555"
      );

      await act(async () => {
        fireEvent.click(screen.getByRole("button"));
      });

      expect(mockChangeStep).toBeCalled();
    });

    it("should have proper error message when input is incorrect", async () => {
      const { result } = renderHook(() => useForm(global.getUseFormProps()));
      const { rerender } = render(
        <YourInfo
          trigger={result.current.trigger}
          register={result.current.register}
          getValues={result.current.getValues}
          errors={result.current.formState.errors}
          changeStep={mockChangeStep}
        />
      );

      fireEvent.input(screen.getByRole("textbox", { name: "Email Address" }), {
        target: { value: "Some Address" }
      });

      await act(async () => {
        await result.current.trigger(["name", "email", "phone"]);
      });

      rerender(
        <YourInfo
          trigger={result.current.trigger}
          register={result.current.register}
          getValues={result.current.getValues}
          errors={result.current.formState.errors}
          changeStep={mockChangeStep}
        />
      );

      expect(
        screen.getByLabelText("Name is a required field!")
      ).toBeInTheDocument();
      expect(
        screen.getByLabelText("Must be a valid email!")
      ).toBeInTheDocument();
      expect(
        screen.getByLabelText("Phone is a required field!")
      ).toBeInTheDocument();

      await act(async () => {
        fireEvent.click(screen.getByRole("button"));
      });

      expect(mockChangeStep).not.toBeCalled();
    });
  });
});
