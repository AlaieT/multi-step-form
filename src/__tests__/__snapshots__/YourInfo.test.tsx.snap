// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YourInfo correct render should match snapshot 1`] = `
<DocumentFragment>
  <div
    id="step"
  >
    <div
      id="content"
    >
      <div
        id="header"
      >
        <h1>
          Personal info
        </h1>
        <h2>
          Please provide your name, email address, and phone number.
        </h2>
      </div>
      <div
        id="fields"
      >
        <div
          class="input field"
        >
          <div
            class="badges"
          >
            <label
              for="name"
            >
              Name
            </label>
          </div>
          <input
            id="name"
            name="name"
            placeholder="e.g. Stephen King"
            type="text"
            value=""
          />
        </div>
        <div
          class="input field"
        >
          <div
            class="badges"
          >
            <label
              for="email"
            >
              Email Address
            </label>
          </div>
          <input
            id="email"
            name="email"
            placeholder="e.g. stephenking@lorem.com"
            type="email"
            value=""
          />
        </div>
        <div
          class="input field"
        >
          <div
            class="badges"
          >
            <label
              for="phone-number"
            >
              Phone Number
            </label>
          </div>
          <input
            id="phone-number"
            name="phone"
            placeholder="e.q. +1 234 567 890"
            type="tel"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      id="controls"
    >
      <button
        class="button"
        id="next_step"
        type="button"
      >
        Next Step
      </button>
    </div>
  </div>
</DocumentFragment>
`;
